hashicorp/consul
# Add support for defining Vault address and token through environment variables in Consul Connect CA provider

### Description

- Add support for defining the Vault token from the `CONSUL_CA_VAULT_TOKEN` environment variable for the Consul Connect CA provider. If the token is provided through the environment variable, it takes precedence over the value set in the configuration file.

- Add support for defining Vault address from the `CONSUL_CA_VAULT_ADDR` environment variable for the Consul Connect CA provider.  If the address is provided through the environment variable, it takes precedence over the value set in the configuration file.

This PR also adds validation and tests for the Vault address. Previously it was possible to start the provider with an empty address, leading to runtime errors such as:

```
[ERROR] connect.ca: Failed to initialize Connect CA: routine="CA initialization" error="error configuring provider: Get \"/v1/auth/token/lookup-self\": unsupported protocol scheme \"\""
```

The parameter is marked as a required one already in the provider documentation.

### Testing & Reproduction steps

- Follow the [Connect CA getting started guide](https://developer.hashicorp.com/consul/tutorials/vault-secure/vault-pki-consul-connect-ca)
- Set the Vault token through `CONSUL_CA_VAULT_TOKEN` environment variable for the Consul process.
- Comment out the `token` field from the Consul configuration file upon testing.
- Verify that Consul Connect CA provider is able to interact with Vault with the token set through the environment variable.

Same test for `CONSUL_CA_VAULT_ADDR` environment variable and the `address` config file parameter.

### Links

Fixes #16853. 

### PR Checklist
